api/v1alpha1/kongplugin_installation_types.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2022 Kong Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	SchemeBuilder.Register(&KongPluginInstallation{}, &KongPluginInstallationList{})
+}
+
+//+genclient
+//+k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:shortName=kpi,categories=kong;all
+//+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name="Ready",type=string,JSONPath=`.status.conditions[?(@.type=="Ready")].status`
+
+// KongPluginInstallation allows using a custom Kong Plugin distributed as a container image available in a registry.
+// Such a plugin can be associated with GatewayConfiguration or DataPlane to be available for particular Kong Gateway
+// and configured with KongPlugin CRD.
```

Comment:
I think that name + namespace should be enough. Because as you mentioned I can't specify new fields in GWAPI `Gateway`. So in the design doc CRDs `GatewayConfiguration` and `DataPlane` are places to add additional fields. It should serve all purposes. And there is one place to influence Gateway's configuration a `GatewayConfiguration` CRD.

The idea with annotations is OK too, but harder to extend in the future. I can't say now whether we would extend something related to this... For me, annotations are something that you use when you can't modify CRD and you want to configure it and there is no other way to achieve it. But it's more I think personal/community opinion/preference

I thought that in case of want a different image (a different version) a separate instance of `KongPluginInstallation` would be configured. 

In terms of docstrings, I think we can adjust them when we sort out the above

The open question is that maybe instead of `KongPluginInstallation` it should be called `KongPluginInstance` as @mlavacca proposed in the design doc.

Reply:
Harder to extend in the future, yeah, but unless we can think of something obvious it's YAGNI territory. If we do find that we need something more complex in the binding, worst case we have two methods for binding, simple and more complex.

Sync discussion came down on the side of only allowing one DataPlane or Gateway to bind a KPI, so if we go with GatewayConfiguration, that single-attach relationship would presumably extend to the whole GatewayConfiguration CRD. That may make sense anyway, but AFAIK it's not something we've enforced in the past.

I like having config visibility directly on the Gateway resource itself when possible (though I recognize we've already gone well past that point with everything else in GatewayConfiguration) and having the same interface for both DataPlane and Gateway, but similar fields in a spec is probably close enough for it to not matter that much.

I prefer `KongPluginInstallation` since it's more descriptive (it's a CR that installs plugins, a name that reflects that is good).

Reply:
To conclude, the decision has been made
- name `KongPluginInstallation`
- `KongPluginInstallation` can be configured in `GatewayConfiguration` and `DataPlane`